Adding a layout for Ensoniq ESQ-1.

felipesanches opened the pull request:
- It can play notes by clicking on the keyboard.
- Volume and Data Entry sliders are hooked up via Lua scripting.
- Volume slider still non-functional. Data Entry slider works fine.
- SVG snippets were kept to the absolute minimum (drawings of envelope and oscillators diagrams)
- The full unit was drawn, but only the relevant part is visible on the default layout. We may later add other views such as "Full keyboard".

## Before

<img width="2034" height="277" alt="Screenshot From 2025-08-12 19-01-09" src="https://github.com/user-attachments/assets/ea614e17-bc54-49c5-a581-1377a848a55a" />


## After

<img width="2034" height="1115" alt="Screenshot From 2025-08-12 18-53-10" src="https://github.com/user-attachments/assets/ab43ad0f-c733-49b9-b695-e756ad9fb765" />

galibert commented on src/mame/ensoniq/esq1.cpp:
```diff
@@ -665,32 +693,151 @@ void esq1_state::sq80(machine_config &config)
+	PORT_BIT(0xf8, IP_ACTIVE_HIGH, IPT_UNUSED)
+
+	PORT_START("INTERNAL_KBD0")
+	PORT_BIT(0x001, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CHANGED_MEMBER(DEVICE_SELF, FUNC(esq1_state::internal_kbd_key_stroke), 0x98) PORT_NAME("C0")
```
Can you use the PORT_GM_* stuff for the key names?  It's not adding anything right now, true, but I hope it will in the future.

galibert replied:
Using (usb-)midi as an input device is not implemented yet.  PORT_GM_* is one of the steps in that direction though, which is why I try to have it comprehensive in the drivers, even if right now it's only setting the name.

Right now when you play on your m-wave, it goes to the esq1 midi in indeed.  It's not seen as a piano keyboard press.

galibert replied:
It replaces the PORT_NAME

felipesanches commented:
The octaves are wrong! I'll reassign the midi note values to fit the correct range.

happppp commented on src/mame/layout/esq1.lay:
```diff
@@ -0,0 +1,708 @@
+	<element name="audio_out_text"><text string="AUDIO OUT"><color red="0.83" green="0.86" blue="0.83" /></text></element>
+
+
+	<!-- TODO: SVG images for the logos
+		<element name="ensoniq_logo"><text string="ensoniq"><color red="0.83" green="0.86" blue="0.83" /></text></element>
+	-->
```
Trademark issue anyway if MAME adds it in the source, and claims CC0.

felipesanches replied:
Cool. But could we leave the comment as a tip to external layout authors?

felipesanches replied:
> The version of the Ensoniq wordmark from this era expired long ago. Creative has kept a different Ensoniq wordmark active (last renewed 2016), but it's visually different and the last Ensoniq branded anything was 2002. So I'm not worried about trademark here.
> 
> That said, the human-readable SVG thing is real, so this isn't a TODO. I don't have an issue with it being a comment, just having it marked TODO, which has a specific meaning in MAME.

Well... I think I can make it fairly readable:

```xml
<svg width="286" height="70">
	<g fill="#e2dbce" stroke="none">
		<!-- E -->
		<path d="
			M 0,0 v 20 h 67 v -20 Z
			m 0,25 v 20 h 67 v -20 z
			m 0,25 v 20 h 67 v -20 z"/>

		<!-- S -->
		<path d="
			M 76,0
			v  45 h  47 v  5 h -47 v  20 h 67
			v -45 h -47 v -5 h  47 v -20 Z" />

		<!-- Q -->
		<path d="
			M 152,0
			v 70 h 86 v -20 h -19 V 0 Z
			m 20,20 h 27 v 30 h -27 z" />

		<!-- 1 -->
		<path d="
			M 266,0
			l -12,20 h 12 v 50 h 20 V 0 Z" />
	</g>
	
	<!-- The triangle -->
	<path
		fill="#36dbe9" stroke="none"
		d="m 222,0 h 39 l -19.5,32.5 z"/>
</svg>
```

<img width="1665" height="487" alt="Screenshot From 2025-08-15 15-33-11" src="https://github.com/user-attachments/assets/cd0a44cc-65ee-4b88-936e-aaf2795a8bb8" />

felipesanches commented on src/mame/layout/esq1.lay:
```diff
@@ -0,0 +1,708 @@
+		<rect state="1"><color red="0.41" green="0.43" blue="0.41"/></rect>
+	</element>
+
+	<element name="white_rect">
+		<rect><color red="0.83" green="0.86" blue="0.83" /></rect>
+	</element>
+
+	<element name="grey_rect">
+		<rect><color red="0.2" green="0.2" blue="0.2" /></rect>
+	</element>
+
+	<element name="dark_grey_rect">
+		<rect><color red="0.13" green="0.13" blue="0.15" /></rect>
+	</element>
+
+	<element name="panel_sheet">
+		<rect><color red="0.17" green="0.17" blue="0.18"/></rect>
+	</element>
+
+	<element name="screen_border">
+		<rect><color red="0.05" green="0.05" blue="0.05"/></rect>
+	</element>
+
+	<element name="invisible-rect">
+		<rect><color red="0" green="0" blue="0" alpha="0"/></rect>
+	</element>
+
+	<element name="black-rect">
+		<rect><color red="0.0" green="0.0" blue="0.0"/></rect>
+	</element>
```
yeah, my rationale was exactly what @happppp described here. The handling of colors in the .lay schema is quite insane. In the near future I'd like to propose improvements for a version 3 with better ways to author layouts and better color declarations are one of the things I would include.

felipesanches commented on src/mame/layout/esq1.lay:
```diff
@@ -0,0 +1,708 @@
+			<bounds state="100" x="9" y="11" width="29" height="16"/>
+			<bounds state="0" x="9" y="66" width="29" height="16"/>
+		</element>
+		<element ref="white_rect"><bounds x="0" y="20" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="0" y="29" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="0" y="38" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="0" y="47" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="0" y="56" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="0" y="65" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="0" y="74" width="9" height="1"/></element>
+
+		<element ref="white_rect"><bounds x="38" y="20" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="38" y="29" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="38" y="38" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="38" y="47" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="39" y="56" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="38" y="65" width="9" height="1"/></element>
+		<element ref="white_rect"><bounds x="38" y="74" width="9" height="1"/></element>
```
Would using `<repeat>` here help? Or would it still be the same issue?

>> Remember MAME still thinks immediate mode is cool, so this overhead adds up quickly.

Maybe it is clear to you because you know about some internal implementation detail, but I fail to comprehend the meaning of this sentence. What is the thing you refer to when you use this "immediate mode" expression?

felipesanches replied:
Let me take the opportunity to mention that I think sliders should be supported natively by the layout system. The current approach of building our own slider all the time by copy-pasting the same lua script everywhere we need it is a mess. It is very error prone and it is not a joy for layout authors.

This demonstrates that there's a need for the feature. And also that it would be useful to have a way to include external scripts and then we could maintain a small library of useful reusable things (like the slider script and also external assets such as SVG files - as described by @rb6502 and others at https://github.com/mamedev/mame/pull/12735#issuecomment-2590291368).

galibert replied:
Believe us, we know.  It is, as is often the case, more complicated that it should be to add.

felipesanches replied:
> For reference, "immediate mode" means

Thanks for the clarification!

felipesanches replied:
@cuavas, does https://github.com/mamedev/mame/pull/14049/commits/b511c53957f006715ed3ac1a3b10155ce04e2ef3 properly address your concern?

felipesanches commented on src/mame/layout/esq1.lay:
```diff
@@ -0,0 +1,708 @@
+	<element ref="dark_grey_rect"><bounds x="-490" y="-95" width="27" height="620"/></element> <!-- left border -->
+	<element ref="dark_grey_rect"><bounds x="1320" y="-95" width="27" height="620"/></element> <!-- right border -->
+	<element ref="dark_grey_rect"><bounds x="-390" y="310" width="24" height="145"/></element> <!-- pitch wheel -->
+	<element ref="dark_grey_rect"><bounds x="-315" y="310" width="24" height="145"/></element> <!-- mod wheel -->
```
done

felipesanches commented on src/mame/layout/esq1.lay:
```diff
@@ -0,0 +1,708 @@
+			   m -57,0 h -9 l -25,-25 h -92 l 0,28 h 176" />
+		</svg>
+		]]></data></image>
+	</element>
```
done

felipesanches commented:
I've just fixed the handling of the underline display command and updated the layout to render it properly. Finally we have visual feedback for the user indicating which is the currently selected item in the ESQ-1 VFD.

Here, the selected voice is "HARP" on internal instrument bank number 3:
<img width="1402" height="281" alt="image" src="https://github.com/user-attachments/assets/add2f70b-5ee5-415e-b0bf-d6b2020e4431" />

And here the waveform setting ("PULSE") is selected for oscillator number 2:
<img width="1402" height="281" alt="Screenshot From 2025-08-19 03-48-32" src="https://github.com/user-attachments/assets/038c2c61-0eb6-403f-8cf2-0174d51ae339" />

The underlines are also correctly rendered on the VFD-only view:
<img width="2034" height="355" alt="image" src="https://github.com/user-attachments/assets/1b90e640-895d-4647-ae3e-ff92b0a2a549" />

felipesanches commented:
Adopted suggestion by @happppp to use SVG to draw the up & down arrows instead of relying on the font having glyphs for these very specific unicode codepoints.  

# Before/After:

<img width="178" height="224" alt="Screenshot From 2025-08-19 12-58-52" src="https://github.com/user-attachments/assets/87cead3b-ef58-484b-888c-d633ea74ab48" />
<img width="178" height="224" alt="Screenshot From 2025-08-19 12-59-37" src="https://github.com/user-attachments/assets/86c20688-3687-4f12-8b07-87d0282224da" />

felipesanches commented:
Also improved the rendering of arrows in the oscillators diagram, by having them filled, as in the original panel artwork.

# Before/After:

<img width="178" height="224" alt="Screenshot From 2025-08-19 12-58-33" src="https://github.com/user-attachments/assets/334c1514-c685-42d4-a1c4-191cbeb8d103" />
<img width="178" height="224" alt="Screenshot From 2025-08-19 12-57-46" src="https://github.com/user-attachments/assets/7a2f5d41-1fa9-4f6f-8c42-6fce9708d57c" />

felipesanches commented:
Very minor tweak here. Better shapes for the arrows in the oscillators diagram.

# Before / After
<img width="329" height="273" alt="Screenshot From 2025-08-21 19-46-22" src="https://github.com/user-attachments/assets/de9d2b62-3422-4d18-8f36-6bf2df5dd720" />
<img width="329" height="273" alt="Screenshot From 2025-08-21 19-41-57" src="https://github.com/user-attachments/assets/49f25271-383a-496f-8abd-347d790a3533" />

cuavas commented on src/mame/layout/esq1.lay:
```diff
@@ -0,0 +1,895 @@
+		</rect>
+	</element>
+
+	<element name="slider_ticks">
+		<image><data><![CDATA[
+			<svg width="46" height="93">
+				<path fill="none" stroke="#d4dad3" stroke-width="1"
+					d="
+						M  0,18 h 9 m 28,0 h 9
+						m -46,9 h 9 m 28,0 h 9
+						m -46,9 h 9 m 28,0 h 9
+						m -46,9 h 9 m 28,0 h 9
+						m -46,9 h 9 m 28,0 h 9
+						m -46,9 h 9 m 28,0 h 9
+						m -46,9 h 9 m 28,0 h 9
+					" />
+			</svg>
+		]]></data></image>
+	</element>
+
+	<group name="slider">
+		<element ref="black-rect">  <!-- slider channel -->
+			<bounds x="12" y="0" width="23" height="94"/>
+		</element>
+		<element ref="dark_grey_rect">  <!-- slider rail -->
+			<bounds x="20" y="10" width="7" height="77"/>
+		</element>
+		<element id="slider_~slider_id~" ref="invisible-rect">
+			<bounds x="12" y="10" width="23" height="77"/>
+		</element>
+		<element ref="slider_ticks"><bounds x="0" y="0" width="46" height="94"/></element>
+		<element id="slider_knob_~slider_id~" ref="slider-knob">
+			<animate inputtag="~port_name~" inputmask="0x7f"/>
+			<bounds state="100" x="9" y="11" width="29" height="16"/>
+			<bounds state="0" x="9" y="66" width="29" height="16"/>
+		</element>
+	</group>
```
You don’t necessarily need to use SVG for the ticks – you could use a bunch of `<rect>` components in the element.

But can you combine the channel, rail and ticks into a single element?  You should be able to just stack those components within the element.  That will give one quad rather three to composite.

felipesanches replied:
It was a bunch of `<rect>` components before. I changed it to a single element [per your request](https://github.com/mamedev/mame/pull/14049#discussion_r2273917026) (and I decided to use a single `<svg>` element to make it nice and clean).

I'm having a hard time understanding your request here. Isn't stacking a group of elements what a `<group>` does?

In cases like this one when it is not yet clear what a request means even after you requested it twice, it would be helpful if you could actually type a snippet of code as an example of what you want to see.

Also, your wording was "to reduce the amount of geometry you’re pushing around", but the only thing that is dragged is the slider knob. All the rest remains static.

Have you missed [this question](https://github.com/mamedev/mame/pull/14049#discussion_r2279750841) about ways to measure these hypothetical performance issues?

Finally, this style of slider implementation that I am using here is almost identical to the reference code I copied from (`src/mame/layout/oberheim_dmx.lay`). Does that implementation also have an issue to be addressed, then?

And have you seen [this question](https://github.com/mamedev/mame/pull/14049#issuecomment-3212443497) about how to animate spring behaviour of a mod wheel?

cuavas reviewed (commented):
This is looking quite a bit better than last time.

After what @happppp said, I’m still not sure about including logos in internal artwork.

felipesanches commented:
> This is looking quite a bit better than last time.
> 
> After what @happppp said, I’m still not sure about including logos in internal artwork.

The svg of the Ensoniq logo that I hand optimized in this PR came from here https://commons.wikimedia.org/wiki/File:Ensoniq_logo.svg where it's legal status is described as:

> The depicted text is ineligible for [copyright](https://en.wikipedia.org/wiki/Copyrights) and therefore in the [public domain](https://en.wikipedia.org/wiki/public_domain) because it is not a “literary work” or other protected type in sense of the local copyright law. Facts, data, and unoriginal information which is common property without sufficiently creative authorship in a general typeface or basic handwriting, and simple geometric shapes are not protected by copyright.

And regarding trademark law, @rb6502 said (at https://github.com/mamedev/mame/pull/14049#discussion_r2279117011) that "The version of the Ensoniq wordmark from this era expired long ago."

happppp commented:
Here is the trademark: https://trademarks.justia.com/735/24/ensoniq-73524524.html
Status: 800 - Registered And Renewed
Owner: CREATIVE TECHNOLOGY LTD.

felipesanches commented:
> Here is the trademark: https://trademarks.justia.com/735/24/ensoniq-73524524.html Status: 800 - Registered And Renewed Owner: CREATIVE TECHNOLOGY LTD.
> 
> And yes, trademark != copyright. But that doesn't mean the trademarks are public domain, they're still intellectual property.

OK. Thanks for the clarification. I will remove the commit that included the Ensoniq logo.

What about the ESQ-1 logo? Is it legally safe to keep? Or should it also be removed?

happppp commented:
No idea, I'd say better be safe than sorry and don't copy-paste brand logos.

rb6502 commented:
That's not the same version of the wordmark that the ESQ-1 used.  That specific version is listed as abandoned.

felipesanches commented:
I have just force-pushed after removing the commits with the brand logos.

The fully branded version is now hosted as an external layout at my personal repo at https://github.com/ArqueologiaDigital/MAME_synths_artwork for anyone who may be interested in using it.

cuavas commented:
> I have just force-pushed after removing the commits with the brand logos.
> 
> The fully branded version is now hosted as an external layout at my personal repo at https://github.com/ArqueologiaDigital/MAME_synths_artwork for anyone who may be interested in using it.

Thanks.  I’d prefer to err on the side of not doing things that could be construed as implying endorsement.

cuavas commented:
> The repo at https://github.com/ArqueologiaDigital/MAME_synths_artwork is completely mine, with no endorsement from MAME devs. Sorry if that gave a bad impression. I am personally completely responsible for what I do in that repo.

I was thanking you for removing it from the PR, that’s all.
